In [ ]:
!python --version

Python 3.12.12


#**Cleaning the Dataset for Housing Prices Prediction**

---

**Objective:**

* Clean the dataset of housing prices containing variables describing different aspects of the houses.

* Remove the variables with the lowest predictive value among the 79 explanatory variables.

* Prepare the dataset for a exploratory data analysis and baseline linear regression.

**Dataset:**

The dataset used in this project is the Ames Housing dataset which was compiled by Dean De Cock. The version used here is downloaded from the [Kaggle House Prices - Advanced Regression](https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/overview) competition page.





###**1. Importing Libraries and Loading the Dataset**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

####**1.1. Loading the Dataset:**

In [ ]:
from google.colab import files

uploaded = files.upload()

df_train = pd.read_csv('train.csv')


Saving train.csv to train.csv


####**1.2. Data Inspection:**

 Use the **df.shape**, **df.info**, **df.describe** methods to see the structure of the dataset.

 To further understand the dataset load and print the metadata. (Note: I have moved the `data_description.txt` to the bottom as an *appendix* to the notebook to avoid derailing the notebook structure.)

In [ ]:
pd.set_option('display.max_columns', None)
df_train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [ ]:
df_train.shape

(1460, 81)

In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [ ]:
df_train.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726   
std       1.112799    30.202904     20.645407   181.066207   456.098091   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000   

        BsmtFinSF2    BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  \
count  1460.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean     46.549315   567.240411  1057.429452  1162.626712   346.992466   
std     161.319273   441.866955   438.705324   386.587738   436.528436   
min       0.000000     0.000000     0.000000   334.000000     0.000000   
25%       0.000000   223.000000   795.750000   882.000000     0.000000   
50%       0.000000   477.500000   991.500000  1087.000000     0.000000   
75%       0.000000   808.000000  1298.250000  1391.250000   728.000000   
max    1474.000000  2336.000000  6110.000000  4692.000000  2065.000000   

       LowQualFinSF    GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath  \
count   1460.000000  1460.000000   1460.000000   1460.000000  1460.000000   
mean       5.844521  1515.463699      0.425342      0.057534     1.565068   
std       48.623081   525.480383      0.518911      0.238753     0.550916   
min        0.000000   334.000000      0.000000      0.000000     0.000000   
25%        0.000000  1129.500000      0.000000      0.000000     1.000000   
50%        0.000000  1464.000000      0.000000      0.000000     2.000000   
75%        0.000000  1776.750000      1.000000      0.000000     2.000000   
max      572.000000  5642.000000      3.000000      2.000000     3.000000   

          HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  \
count  1460.000000   1460.000000   1460.000000   1460.000000  1460.000000   
mean      0.382877      2.866438      1.046575      6.517808     0.613014   
std       0.502885      0.815778      0.220338      1.625393     0.644666   
min       0.000000      0.000000      0.000000      2.000000     0.000000   
25%       0.000000      2.000000      1.000000      5.000000     0.000000   
50%       0.000000      3.000000      1.000000      6.000000     1.000000   
75%       1.000000      3.000000      1.000000      7.000000     1.000000   
max       2.000000      8.000000      3.000000     14.000000     3.000000   

       GarageYrBlt   GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  \
count  1379.000000  1460.000000  1460.000000  1460.000000  1460.000000   
mean   1978.506164     1.767123   472.980137    94.244521    46.660274   
std      24.689725     0.747315   213.804841   125.338794    66.256028   
min    1900.000000     0.000000     0.000000     0.000000     0.000000   
25%    1961.000000     1.000000   334.500000     0.000000     0.000000   
50%    1980.000000     2.000000   480.000000     0.000000    25.000000   
75%    2002.000000     2.000000   576.000000   168.000000    68.000000   
max    2

###**2. Data Cleaning and Wrangling:**

---
Clean the data by dropping unnecessary columns and imputing missing values.




####**2.1. Dropping Columns with majority missing values:**

In [ ]:
housing_train = df_train.copy()

housing_train.drop(columns=['Id'], inplace=True)
housing_train.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd        NaN   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng        NaN   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd             7        Typ           

Check which columns have missing values:

In [ ]:
missing_counts = housing_train.isna().sum()

for column, count in missing_counts.items():
  if count > 0:
    print(column, count)

LotFrontage 259
Alley 1369
MasVnrType 872
MasVnrArea 8
BsmtQual 37
BsmtCond 37
BsmtExposure 38
BsmtFinType1 37
BsmtFinType2 38
Electrical 1
FireplaceQu 690
GarageType 81
GarageYrBlt 81
GarageFinish 81
GarageQual 81
GarageCond 81
PoolQC 1453
Fence 1179
MiscFeature 1406


Drop columns with 90% and above missing data since any imputation will have very unreliable results:

In [ ]:
housing_train.drop(columns = ["Alley", "MiscFeature", "Fence"], inplace=True)
housing_train.shape

(1460, 77)

Despite three digit missing values we will keep `LotFrontage`, `PoolQC` and `FirepalceQu` because it is still possible to impute them by extrapolating from related columns. Also, missing ≠ unimportant.

In [ ]:
missing_counts = housing_train.isna().sum()

for column, count in missing_counts.items():
  if count > 0:
    print(column, count)

LotFrontage 259
MasVnrType 872
MasVnrArea 8
BsmtQual 37
BsmtCond 37
BsmtExposure 38
BsmtFinType1 37
BsmtFinType2 38
Electrical 1
FireplaceQu 690
GarageType 81
GarageYrBlt 81
GarageFinish 81
GarageQual 81
GarageCond 81
PoolQC 1453


####**2.2. Imputing Missing Values**
---
In this section of the notebook we will impute the missing values and save it to an imputed dataframe.



Impute `Electrical` with the most common value, `SBrkr` which is the mode of the column.

In [ ]:
housing_train['Electrical'].fillna(housing_train['Electrical'].mode()[0], inplace=True)

/tmp/ipython-input-1137988972.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  housing_train['Electrical'].fillna(housing_train['Electrical'].mode()[0], inplace=True)


We will base the value we fill in for `PoolQC` on `PoolArea` because those houses with no data for pool quality are those with no swimming pool installed in the first place!

In [ ]:
housing_train['PoolQC'] = housing_train.apply(lambda x: 'NA' if x['PoolArea'] == 0 else x['PoolQC'], axis=1)

I checked reddit and I found out that houses on the same zoning and same class usually have the same lot frontage so I decided to group the missing by both categories and impute with class median:

In [ ]:
housing_train['LotFrontage'] = housing_train.groupby(['MSSubClass', 'MSZoning'])['LotFrontage'].transform(lambda x: x.fillna(x.median()))

The same with pool quality, fireplace quality is most likely left blank by people who answered 0 to number of fireplaces in their listing. So we impute `FireplaceQu` by assigning *NA* for *No Fireplace* for rows where `Fireplaces` has the value 0.

In [ ]:
housing_train['FireplaceQu'] = housing_train.apply(lambda x: 'NA' if x['Fireplaces'] == 0 else x['FireplaceQu'], axis=1)

Imputing `GarageType` and garage-related columns:

In [ ]:
df_garage = housing_train[['GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond']].copy()
df_garage.head(15)

GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual  \
0      Attchd       2003.0          RFn           2         548         TA   
1      Attchd       1976.0          RFn           2         460         TA   
2      Attchd       2001.0          RFn           2         608         TA   
3      Detchd       1998.0          Unf           3         642         TA   
4      Attchd       2000.0          RFn           3         836         TA   
5      Attchd       1993.0          Unf           2         480         TA   
6      Attchd       2004.0          RFn           2         636         TA   
7      Attchd       1973.0          RFn           2         484         TA   
8      Detchd       1931.0          Unf           2         468         Fa   
9      Attchd       1939.0          RFn           1         205         Gd   
10     Detchd       1965.0          Unf           1         384         TA   
11    BuiltIn       2005.0          Fin           3         736         TA   
12     Detchd       1962.0          Unf           1         352         TA   
13     Attchd       2006.0          RFn           3         840         TA   
14     Attchd       1960.0          RFn           1         352         TA   

   GarageCond  
0          TA  
1          TA  
2          TA  
3          TA  
4          TA  
5          TA  
6          TA  
7          TA  
8          TA  
9          TA  
10         TA  
11         TA  
12         TA  
13         TA  
14         TA

In [ ]:
df_garage.isna().sum()

GarageType      81
GarageYrBlt     81
GarageFinish    81
GarageCars       0
GarageArea       0
GarageQual      81
GarageCond      81
dtype: int64

In [ ]:
df_garage.isna().sum(axis=1).value_counts()

0    1379
5      81
Name: count, dtype: int64

In [ ]:
temp = housing_train[housing_train['GarageType'].isna()][
    ['GarageType','GarageYrBlt', 'GarageFinish', 'GarageQual', 'GarageCond']
]
temp.head(81)

GarageType  GarageYrBlt GarageFinish GarageQual GarageCond
39          NaN          NaN          NaN        NaN        NaN
48          NaN          NaN          NaN        NaN        NaN
78          NaN          NaN          NaN        NaN        NaN
88          NaN          NaN          NaN        NaN        NaN
89          NaN          NaN          NaN        NaN        NaN
...         ...          ...          ...        ...        ...
1349        NaN          NaN          NaN        NaN        NaN
1407        NaN          NaN          NaN        NaN        NaN
1449        NaN          NaN          NaN        NaN        NaN
1450        NaN          NaN          NaN        NaN        NaN
1453        NaN          NaN          NaN        NaN        NaN

[81 rows x 5 columns]

This confirms my suspicion that all the missing values in garage-related data are in the same rows, those that align with no garage built. We impute this by filling the rows in with 'NA' which is the value for no garage according to `data_description.txt`:

In [ ]:
cols = ['GarageType', 'GarageYrBlt', 'GarageFinish','GarageQual', 'GarageCond']
housing_train[cols] = housing_train[cols].fillna('NA')

In [ ]:
missing_counts = housing_train.isna().sum()

for column, count in missing_counts.items():
  if count > 0:
    print(column, count)

MasVnrType 872
MasVnrArea 8
BsmtQual 37
BsmtCond 37
BsmtExposure 38
BsmtFinType1 37
BsmtFinType2 38


Impute the basement-related variables:

In [ ]:
df_basement = housing_train[['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'BsmtFinSF1', 'BsmtFinSF2','BsmtUnfSF', 'TotalBsmtSF']].copy()
df_basement.head(10)

BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2  BsmtFinSF1  \
0       Gd       TA           No          GLQ          Unf         706   
1       Gd       TA           Gd          ALQ          Unf         978   
2       Gd       TA           Mn          GLQ          Unf         486   
3       TA       Gd           No          ALQ          Unf         216   
4       Gd       TA           Av          GLQ          Unf         655   
5       Gd       TA           No          GLQ          Unf         732   
6       Ex       TA           Av          GLQ          Unf        1369   
7       Gd       TA           Mn          ALQ          BLQ         859   
8       TA       TA           No          Unf          Unf           0   
9       TA       TA           No          GLQ          Unf         851   

   BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  
0           0        150          856  
1           0        284         1262  
2           0        434          920  
3           0        540          756  
4           0        490         1145  
5           0         64          796  
6           0        317         1686  
7          32        216         1107  
8           0        952          952  
9           0        140          991

In [ ]:
df_basement.isna().sum()

BsmtQual        37
BsmtCond        37
BsmtExposure    38
BsmtFinType1    37
BsmtFinType2    38
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUnfSF        0
TotalBsmtSF      0
dtype: int64

In [ ]:
temp2 = housing_train[housing_train['BsmtExposure'].isna()][
    ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'BsmtFinSF1', 'BsmtFinSF2','BsmtUnfSF', 'TotalBsmtSF']
]
temp2.head(38)

BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2  BsmtFinSF1  \
17        NaN      NaN          NaN          NaN          NaN           0   
39        NaN      NaN          NaN          NaN          NaN           0   
90        NaN      NaN          NaN          NaN          NaN           0   
102       NaN      NaN          NaN          NaN          NaN           0   
156       NaN      NaN          NaN          NaN          NaN           0   
182       NaN      NaN          NaN          NaN          NaN           0   
259       NaN      NaN          NaN          NaN          NaN           0   
342       NaN      NaN          NaN          NaN          NaN           0   
362       NaN      NaN          NaN          NaN          NaN           0   
371       NaN      NaN          NaN          NaN          NaN           0   
392       NaN      NaN          NaN          NaN          NaN           0   
520       NaN      NaN          NaN          NaN          NaN           0   
532       NaN      NaN          NaN          NaN          NaN           0   
533       NaN      NaN          NaN          NaN          NaN           0   
553       NaN      NaN          NaN          NaN          NaN           0   
646       NaN      NaN          NaN          NaN          NaN           0   
705       NaN      NaN          NaN          NaN          NaN           0   
736       NaN      NaN          NaN          NaN          NaN           0   
749       NaN      NaN          NaN          NaN          NaN           0   
778       NaN      NaN          NaN          NaN          NaN           0   
868       NaN      NaN          NaN          NaN          NaN           0   
894       NaN      NaN          NaN          NaN          NaN           0   
897       NaN      NaN          NaN          NaN          NaN           0   
948        Gd       TA          NaN          Unf          Unf           0   
984       NaN      NaN          NaN          NaN          NaN           0   
1000      NaN      NaN          NaN          NaN          NaN           0   
1011      NaN      NaN          NaN          NaN          NaN           0   
1035      NaN      NaN          NaN          NaN          NaN           0   
1045      NaN      NaN          NaN          NaN          NaN           0   
1048      NaN      NaN          NaN          NaN          NaN           0   
1049      NaN      NaN          NaN          NaN          NaN           0   
1090      NaN      NaN          NaN          NaN          NaN           0   
1179      NaN      NaN          NaN          NaN          NaN           0   
1216      NaN      NaN          NaN          NaN          NaN           0   
1218      NaN      NaN          NaN          NaN          NaN           0   
1232      NaN      NaN          NaN          NaN          NaN           0   
1321      NaN      NaN          NaN          NaN          NaN           0   
1412      NaN      NaN          NaN          NaN          NaN           0   

      BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  
17             0          0            0  
39             0          0            0  
90             0          0            0  
102            0          0            0  
156            0          0            0  
182            0          0            0  
259            0          0            0  
342            0          0            0  
362            0          0            0  
371            0          0            0  
392            0          0            0  
520            0          0            0  
532            0          0            0  
533            0          0            0  
553            0          0            0  
646            0          0            0  
705            0          0            0  
736            0          0            0  
749            0          0            0  
778            0          0            0  
868            0          0            0  
894            0          0            0  
897         

If the total basement area is 0 sq ft. then there's no basement! We will impute these area by setting the value to *NA* for each row:

In [ ]:
cols = ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']

housing_train[cols] = housing_train.apply(
    lambda x: ['NA']*len(cols) if x['TotalBsmtSF'] == 0 else x[cols], axis=1, result_type='expand')

Fill in the missing value for basement exposure in row 984 with the mode for the class:

In [ ]:
exp_mode = housing_train['BsmtExposure'].mode()[0]
housing_train['BsmtExposure'].fillna(exp_mode, inplace=True)

/tmp/ipython-input-178677849.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  housing_train['BsmtExposure'].fillna(exp_mode, inplace=True)


In [ ]:
cols = ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2']
housing_train[cols].isna().sum()

BsmtQual        0
BsmtCond        0
BsmtExposure    0
BsmtFinType1    0
BsmtFinType2    1
dtype: int64

In [ ]:
temp3 = housing_train[housing_train['BsmtFinType2'].isna()][
    ['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'BsmtFinSF1', 'BsmtFinSF2','BsmtUnfSF', 'TotalBsmtSF']
]
temp3.head()

BsmtQual BsmtCond BsmtExposure BsmtFinType1 BsmtFinType2  BsmtFinSF1  \
332       Gd       TA           No          GLQ          NaN        1124   

     BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  
332         479       1603         3206

If there's a square feet value for finished basement `BsmtFinSF2` that means that `BsmtFinType2` is not NA! We impute this by filling in with the mode:

In [ ]:
bsmt2_mode = housing_train['BsmtFinType2'].mode()[0]
housing_train['BsmtFinType2'].fillna(bsmt2_mode, inplace=True)

/tmp/ipython-input-143673002.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  housing_train['BsmtFinType2'].fillna(bsmt2_mode, inplace=True)


Check where masonry veneer area is empty:

In [ ]:
df_masonry = housing_train[['MasVnrType', 'MasVnrArea']].copy()
temp4 = housing_train[housing_train['MasVnrArea'].isna()][
    ['MasVnrType', 'MasVnrArea']
]
temp4.head(8)

MasVnrType  MasVnrArea
234         NaN         NaN
529         NaN         NaN
650         NaN         NaN
936         NaN         NaN
973         NaN         NaN
977         NaN         NaN
1243        NaN         NaN
1278        NaN         NaN

Where masonry veneer type is empty:

In [ ]:
temp5 = housing_train[housing_train['MasVnrType'].isna()][
    ['MasVnrType', 'MasVnrArea']
]
temp5.head()

MasVnrType  MasVnrArea
1        NaN         0.0
3        NaN         0.0
5        NaN         0.0
8        NaN         0.0
9        NaN         0.0

Masonry veneer type is NA on rows where masonry area is 0 since that means no veneer was installed by the homeowners!

In [ ]:
housing_train.loc[housing_train['MasVnrArea'] == 0, 'MasVnrType'] = 'None'

In [ ]:
missing_counts = housing_train.isna().sum()

for column, count in missing_counts.items():
  if count > 0:
    print(column, count)

MasVnrType 13
MasVnrArea 8


In [ ]:
mask_both_nan = housing_train['MasVnrType'].isna() & housing_train['MasVnrArea'].isna()
housing_train.loc[mask_both_nan, 'MasVnrType'] = 'None'
housing_train.loc[mask_both_nan, 'MasVnrArea'] = 0.0

In [ ]:
masvnr_mode = housing_train['MasVnrType'].mode()[0]
mask_type_nan = housing_train['MasVnrType'].isna() & housing_train['MasVnrArea'].notna()
housing_train.loc[mask_type_nan, 'MasVnrType'] = masvnr_mode


In [ ]:
total_missing = housing_train.isna().sum().sum()
print(total_missing)


0


In [ ]:
housing_train.shape

(1460, 77)

In [ ]:
housing_train.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          60       RL         65.0     8450   Pave      Reg         Lvl   
1          20       RL         80.0     9600   Pave      Reg         Lvl   
2          60       RL         68.0    11250   Pave      IR1         Lvl   
3          70       RL         60.0     9550   Pave      IR1         Lvl   
4          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType  \
0    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
1    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam   
2    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam   
3    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam   
4    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam   

  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle  \
0     2Story            7            5       2003          2003     Gable   
1     1Story            6            8       1976          1976     Gable   
2     2Story            7            5       2001          2002     Gable   
3     2Story            7            5       1915          1970     Gable   
4     2Story            8            5       2000          2000     Gable   

  RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond  \
0  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA   
1  CompShg     MetalSd     MetalSd       None         0.0        TA        TA   
2  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA   
3  CompShg     Wd Sdng     Wd Shng       None         0.0        TA        TA   
4  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd        TA   

  Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1  \
0      PConc       Gd       TA           No          GLQ         706   
1     CBlock       Gd       TA           Gd          ALQ         978   
2      PConc       Gd       TA           Mn          GLQ         486   
3     BrkTil       TA       Gd           No          ALQ         216   
4      PConc       Gd       TA           Av          GLQ         655   

  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC  \
0          Unf           0        150          856    GasA        Ex   
1          Unf           0        284         1262    GasA        Ex   
2          Unf           0        434          920    GasA        Ex   
3          Unf           0        540          756    GasA        Gd   
4          Unf           0        490         1145    GasA        Ex   

  CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  \
0          Y      SBrkr       856       854             0       1710   
1          Y      SBrkr      1262         0             0       1262   
2          Y      SBrkr       920       866             0       1786   
3          Y      SBrkr       961       756             0       1717   
4          Y      SBrkr      1145      1053             0       2198   

   BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0             1             0         2         1             3             1   
1             0             1         2         0             3             1   
2             1             0         2         1             3             1   
3             1             0         1         0             3             1   
4             1             0         2         1             4             1   

  KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  \
0          Gd             8        Typ           0          NA     Attchd   
1          TA             6        Typ           1          TA     Attchd   
2          Gd             6        Typ           1          TA     Attchd   
3          Gd             7

###**3. Saving Dataset for Analysis:**

The data has now been cleaned and completed and will now be used for exploratory data analysis and for training the machine learning models!

In [ ]:
housing_train.to_csv('housing_train_cleaned.csv', index=False)

What I wanna do next is group the data into **numerical** (`int64` and `float`) and **categorical** (`object`). Furthermore, I wanna subdivide categorical data into **nominal** or *unordered data* which I can one-hot encode and **ordinal** or *ordered data* which i can map to 1-n scales (like 1-4, 1-5 or 1-7 scales). I will do that in my Exploratory Data Analysis and Feature Engineering Notebook.

###**4. Appendix:**
---
This section shows the structure of the dataset and the definition of each category/column. It also shows the expected row values for each column. I keep this here so I can scroll down to check relationships between columns.

In [ ]:
with open('data_description.txt', 'r') as file:
    metadata = file.read()

print(metadata)

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM